Re-land#2 "Fix thread safety in getCurrentCUDABlasHandle and getCUDABlasLtWorkspace"

t-ivan-gr opened the pull request:
Summary:
getCurrentCUDABlasHandle() and getCUDABlasLtWorkspace() use static mutable maps that are not protected from concurrent read-and-write. This leads to crashes.

This diff adds mutexes to synchronize access to the static maps.

Re-land context:

This is a re-land of https://github.com/pytorch/pytorch/pull/167248.

A few issues were addressed:
- fix for a bug in fast path: premature return in getCurrentCUDABlasHandle)
- fix for test flakiness (https://github.com/pytorch/pytorch/pull/167884)

Test Plan:
1. regression tests:
buck2 test \mode/opt //caffe2/test\:test_transformers_cuda
https://www.internalfb.com/intern/testinfra/testrun/6192449759713581

2. Use a GPU OD, run multi-threaded tests with TSAN:

buck test fbcode//mode/dev-tsan fbcode//caffe2:cuda_cublas_handle_pool_test  -- --stress-runs 100
https://www.internalfb.com/intern/testinfra/testrun/14355223937501118

Differential Revision: D87111985

Skylion007 commented on aten/src/ATen/cuda/CublasHandlePool.cpp:
```diff
@@ -233,6 +241,44 @@ at::DataPtr getNewCUDABlasLtWorkspace() {
+    if (workspace_it == workspace.map.end()) {
+      workspace_it =
+          workspace.map.emplace(key, std::move(new_workspace)).first;
+    }
```
```suggestion
    auto workspace_it = workspace.map.try_emplace(key, std::move(new_workspace)).second.first

t-ivan-gr replied:
Thanks! That makes sense, I applied your suggestions

Skylion007 reviewed (approved):
Slight performance nit. Pretty sure you can use try_emplace here to reduce a secondary lookup of key, speed things up, and make it more concise.

t-ivan-gr commented:
@pytorchbot label "topic: not user facing"

t-ivan-gr committed:
Re-land#2 "Fix thread safety in getCurrentCUDABlasHandle and getCUDABlasLtWorkspace" (#167928)

Summary:


getCurrentCUDABlasHandle() and getCUDABlasLtWorkspace() use static mutable maps that are not protected from concurrent read-and-write. This leads to crashes.

This diff adds mutexes to synchronize access to the static maps.

Re-land context:

This is a re-land of https://github.com/pytorch/pytorch/pull/167248.

A few issues were addressed:
- fix for a bug in fast path: premature return in getCurrentCUDABlasHandle)
- fix for test flakiness (https://github.com/pytorch/pytorch/pull/167884)

Test Plan:
1_. regression tests:
```
buck2 test mode/opt //caffe2/test:test_transformers_cuda
```
https://www.internalfb.com/intern/testinfra/testrun/6192449759713581


2_. Use a GPU OD, run multi-threaded tests with TSAN:

```
buck test fbcode//mode/dev-tsan fbcode//caffe2:cuda_cublas_handle_pool_test  -- --stress-runs 100
```
https://www.internalfb.com/intern/testinfra/testrun/14355223937501118

Differential Revision: D87111985
